crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -535,6 +572,63 @@ def cancel
         upgrade_status.initialize_state
       end
 
+      def upgrade_nodes_disruptive(to_upgrade)
+        to_upgrade.each do |element|
+          # If role has a cluster assigned ugprade that cluster (reuse the non-disruptive code here)
+          if ServiceObject.is_cluster?(element)
+            cluster = ServiceObject.cluster_name element
+            cluster_env = "pacemaker-config-#{cluster}"
+            founder = ::Node.find(
+              "pacemaker_founder:true AND " \
+              "pacemaker_config_environment:#{cluster_env}"
+            ).first
+            upgrade_cluster founder, cluster_env
+          else
+            # if cluster(s) and node(s) assigned to the role are already upgraded, skip to next role
+            # if role has single node(s) assigned upgrade those nodes (optionally all at once)
+            upgrade_one_node element
+            # FIXME: this would be nice to do in paralel...
+          end
+        end
+
+        # FIXME: (maybe) skip cinder-volume nodes here that also have a nova compute role assigned
+        # and handle those as part of the compute node upgrade
+      end
+
+      # Go through active proposals and return elements with assigned roles in the right order
+      def upgradable_elements_of_proposals(proposals)
+        to_upgrade = []
+        # If barclamp has an active proposal go through the roles in element order
+        # (skip the compute roles)
+        proposals.each do |name, propmap|
+          proposal = propmap[:prop]
+          elements = proposal["deployment"][name]["elements"]
+          proposal["deployment"][name]["element_order"].flatten.each do |el|
+            next if el.start_with?("nova-compute") || elements[el].nil?
+            to_upgrade |= elements[el]
+          end
+        end
+        to_upgrade
+      end
+
+      def upgrade_disruptive_controllers
+        save_upgrade_state("Entering disruptive upgrade of controller nodes")
+
+        # Go through all OpenStack barclamps (skip "pacemaker" and "openstack") ordered by run_order
```

Comment:
This is not correct or at least it does not sync with the comment
> Go through all OpenStack barclamps (skip "pacemaker" and "openstack") ordered by run_order

This is not skipping the openstack barclamp but its skipping any barclamps that are NOT in the openstack category, like `ntp` or `provisioner` which have the `crowbar` category. Which Im guessing its the end goal here, so the comment should change.

```
irb(main):021:0> bar = []
irb(main):022:0> BarclampCatalog.barclamps.each do |k, v|
irb(main):023:1* p = Proposal.where(barclamp: k).first
irb(main):024:1> next if p.nil?
irb(main):025:1> next if p.category != "OpenStack"
irb(main):026:1> next if k == "pacemaker"
irb(main):027:1> bar.push(k)
irb(main):028:1> end
irb(main):029:0> bar
=> ["nova", "keystone", "rabbitmq", "neutron", "cinder", "glance", "horizon", "database"]
```